Master merge

tw1tch01 opened the pull request.

tw1tch01 committed:
Initial creation of repo

- commiting all work done so far to repo

tw1tch01 committed:
cache service

- added ICacheService interface layer

tw1tch01 committed:
rename

- renamed CacheKey.Time => CacheKey.Times

tw1tch01 committed:
changes

tw1tch01 committed:
Merge branch 'feature/cache-interface' into develop

tw1tch01 committed:
added model unit tests

tw1tch01 committed:
Added item commands and minor refactoring

- Added item commands
- Added services for item commands
- Added validators for commands
- Added some domain exceptions
- Minor tweeks to TodoItem methods and unit tests

tw1tch01 committed:
Item actions and tests

tw1tch01 committed:
finishing up on item commands

tw1tch01 committed:
Merge branch 'feature/item-commands' into develop

tw1tch01 committed:
note commands

tw1tch01 committed:
missed items in commit

tw1tch01 committed:
fixing merge issues

tw1tch01 committed:
Merge branch 'develop' into feature/note-commands

tw1tch01 committed:
Merge branch 'feature/note-commands' into develop

tw1tch01 committed:
Renamed Models project to DomainModels

- renamed Models project to DomainModels
- fixed all references and namespaces for DomainModels

tw1tch01 committed:
application layer rework

- split application into Application and Services class libraries

tw1tch01 committed:
fixing up tests

tw1tch01 committed:
fixed git ignore

tw1tch01 committed:
git ignore

tw1tch01 committed:
removed obj files

tw1tch01 committed:
notes and tests

tw1tch01 committed:
Merge branch 'feature/layer-rework' into develop

tw1tch01 committed:
upgraded Data

- upgraded Data to fix LinqSpecification operations issue

tw1tch01 committed:
Added notifications

- added notifcations
- added custom mediator to handle notifications in new threads so no waiting is needed
- fix git ignore to include tests and removed obj files

tw1tch01 committed:
item workflows and notifications

tw1tch01 committed:
Added workflow and notifications for notes

- refactored workflow and notifications
- added workflow process and notifications for notes

tw1tch01 committed:
Merge branch 'feature/external-services' into develop

tw1tch01 committed:
cleanup

tw1tch01 committed:
Creating NuGet packages

tw1tch01 committed:
TodoItem validators update

- removed length restriction on descript when creating item
- removed length restriction on descript when updating item

tw1tch01 committed:
resolving dependency injection issues

- added transient injection in Application
- changed singleton to transient injection in Services
- fixed some injections in Services
- updated Application to 1.1.0
- updated Services to 1.1.0

tw1tch01 committed:
Moved classes out of "External" namespace

- moved classes out of "external" namespace
- moved "Events" into its respective folder+namespace in Services
- updated Services to 1.2.0

tw1tch01 committed:
namespace fix

tw1tch01 committed:
internal classes

changed internal class to public with virtual methods

tw1tch01 committed:
fixed dependency injection

- added DI for mapping profile to Services
- updated Services to 1.4.0

tw1tch01 committed:
fixup

tw1tch01 committed:
fixes

tw1tch01 committed:
test fixes

tw1tch01 committed:
Merge branch 'feature/fixup' into develop

tw1tch01 committed:
added domain methods

- added methods that would group the notes on an item with its replies

tw1tch01 committed:
simplied method as values where already available

tw1tch01 committed:
fixed filter by status specification

tw1tch01 committed:
fixed filter by status

tw1tch01 committed:
removed sorting by status

tw1tch01 committed:
Added README.md

- updated Data

tw1tch01 committed:
updated README.md

tw1tch01 committed:
Update README.md

tw1tch01 committed:
Update README.md

ForBalah commented on Src/Application/Services/TodoItems/CachedItemCommands/CachedItemsCommandService.cs:
```diff
@@ -0,0 +1,76 @@
+        public async Task<Guid> CreateItem(CreateItemDto itemDto)
+        {
+            return await _commandService.CreateItem(itemDto);
+        }
```
Rather go
```csharp
public Task<Guid> CreateItem(CreateItemDto itemDto) =>
  _commandService.CreateItem(itemDto);
```
i.e. drop the async and await because you don't need the result of the async within the method, so there's no need to add it. (did you know that every addition of `async` and `await` adds about 100 bytes of code?)

ForBalah commented on Src/Application/Services/TodoItems/CachedItemQueries/CachedItemsQueryService.cs:
```diff
@@ -0,0 +1,54 @@
+            {
+                return await _queryService.GetItem(parentId);
+            });
+        }
```
Once you've mastered the art of removing `async/await` from single-line methods so they only return a task. You can start taking it a step further like this:
```csharp
public Task<TodoItemDetails> GetItem(Guid parentId) =>
  _cacheService.Get(
    CacheKeys.Items.GetItem(parentId),
    CacheKeys.Times.Default,
    () => _queryService.GetItem(parentId));
```
And if you name your parameters, even better!

ForBalah commented on Src/Application/Services/TodoItems/CachedItemCommands/ICachedItemsCommandsService.cs:
```diff
@@ -0,0 +1,8 @@
+{
+    public interface ICachedItemsCommandsService : IItemsCommandService
+    {
+    }
```
I'm guessing you added this (and the other ICached... services) because of the finickiness with dependency injection? I'm going to assume "yes". Empty interfaces are useful if you are using the presence of the interface to drive some logic. Having an implementation of an empty interface isn't a great design - you're breaking the interface segregation principle by needing to implement everything in the "base" interface.

I think what you're looking for here is the "decorator pattern". See the `CachedItemsCommandService` for more.

ForBalah commented on Src/Application/Services/TodoItems/CachedItemCommands/CachedItemsCommandService.cs:
```diff
@@ -0,0 +1,76 @@
+
+namespace Todo.Application.Services.TodoItems.CachedItemCommands
+{
+    public class CachedItemsCommandService : ICachedItemsCommandsService
```
(continuation from `ICachedItemsCommandsService` comment) The decorator pattern says: "I have a class with this behaviour and I want to "decorate" it with additional features. So you're on the right track with this classes constructor taking in an `IItemsCommandService` and you are truly decorating it with caching abilities. But there are 2 missing steps...

step 1: this class should simply inherit from `IItemsCommandService` - your code will doesn't (shouldn't) actually care that it is a cached service. After all, it calls the methods which come from that interface and there are no cache-specific public methods...

setp 2: see `DependencyInjection` class for more...

ForBalah commented on Src/Application/DependencyInjection.cs:
```diff
@@ -0,0 +1,26 @@
+            services.AddTransient<IItemsQueryService, ItemsQueryService>();
+            services.AddTransient<INotesCommandService, NotesCommandService>();
+            services.AddTransient<ICachedItemsCommandsService, CachedItemsCommandService>();
+            services.AddTransient<ICachedItemsQueryService, CachedItemsQueryService>();
```
(continued from `CachedItemsCommandService` comments) You wouldn't want to expose an `ICached...Service` because you don't want your code to have to know what choice to make, especially considering that these inherit from their "base" interface and are empty. So you'll want to use a decorator - have one interface, but you control the behaviour of that interface here, by "decorating" it with the desired implementation.

However, .NET Core's DI is great, but rudimentary (80/20 rule) and decoration falls into that 20%. But! there are ways around it. See [this post](https://andrewlock.net/adding-decorated-classes-to-the-asp.net-core-di-container-using-scrutor/) by Andrew Lock on how to put in the decorator manually, or with a library.

What should the end result look like? well... something like this:
```csharp
// In DependencyInjection
services.AddTransient<IItemsCommandService, ItemsCommandService>();
services.Decorate<IItemsCommandService, CachedItemsCommandService>();

// In code:
public MyController(IItemsCommandService itemsCommand)
{
  // we can now use IItemsCommandService without caring whether there's a caching framework behind it or not.
  ...
}
```

ForBalah commented on Src/Application/Services/TodoItems/ItemQueries/IItemsQueryService.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Todo.Application.Services.TodoItems.ItemQueries
+{
+    public interface IItemsQueryService : IGetItemService,
+                                          IChildItemsLookupService,
+                                          IParentItemsLookupService
+    {
+    }
```
Be wary of interfaces like this - inheritance has its place. But this straddles the open/close principle border - its not obvious but if this needs to be used elsewhere, all of a sudden it has these 3 interfaces' baggage that it has to lug around as well. You'd rather want to use "composition" instead.

Composition says "I want to have a class that has x, y, z behaviour. But instead of inheriting from all those 3, I build up my class to understand x, y, z. That way if I need to replace "x" with "a" in a different setting, I don't have to change my inheritance." How can you do it here? see [this Medium article](https://medium.com/jexia/developer-beware-consider-composition-over-inheritance-a7d4e14556c5) for a summary, and [this wikipedia article](https://en.wikipedia.org/wiki/Composition_over_inheritance) for a rudimentary C# example.

tw1tch01 replied:
Definitely makes sense, however, this interface is used within the decorator implementation... so I'm assuming it would be better to remove this interface and convert the `ItemsQueryService.cs` class into a composition of the query services, however, create cache decorator services for each query service? 🤔

tw1tch01 replied:
Actually! I just realised that is exactly what I should be doing... So creating cached services per query/command service is better, as it satisfies the SRP and OCP, and not only that, if there was a particular query in which I didn't want to cache, I could so easily as I wouldn't "decorate" that query service! Also, caching is more of an infrastructure layer concern, NOT a Core layer concern! Thus, all caching implementations/decorators should be in the infrastructure layer \o/

ForBalah commented on Src/Domain/Common/ICreatedAudit.cs:
```diff
@@ -0,0 +1,11 @@
+    public interface ICreatedAudit
+    {
+        string CreatedBy { get; }
+        DateTime CreatedOn { get; }
```
Consider using DateTimeOffset so that you can more easily see when something was created.

ForBalah commented on Src/Domain/Exceptions/ItemAlreadyStartedException.cs:
```diff
@@ -0,0 +1,19 @@
+            : base(string.Format(_message, cancelledOn, itemId))
+        {
+            CancelledOn = cancelledOn;
+            ItemId = itemId;
```
Consider using the [`Exception.Data` property](https://docs.microsoft.com/en-us/dotnet/api/system.exception.data?view=netframework-4.8) to store this info - it gives you serialization for free!

ForBalah commented on Src/DomainModels/TodoItems/TodoItemLookupParams.cs:
```diff
@@ -0,0 +1,31 @@
+        {
+            return $"{nameof(CreatedAfter)}={CreatedAfter}&" +
+                   $"{nameof(CreatedBefore)}={CreatedBefore}&" +
+                   $"{nameof(SearchBy)}={SearchBy}&" +
```
Since `SearchBy` is a string, you'd want to make doubly sure that your query string is URL escaped properly (otherwise the likelihood of strange errors is increased).

tw1tch01 committed:
pull request changes

Initial pull request changes

tw1tch01 committed:
Services rework

- Added ValidationResult for each service call
- Removed throwing exceptions from service calls, instead returns appropriate ValidationResult
- Updated UnitTests
- Removed Exceptions

tw1tch01 committed:
assembly updates

tw1tch01 committed:
fixes

tw1tch01 committed:
removed factories
